### DISCUSSION TITLE
AsyncSession. AsyncResult.first() freezes asyncio loop

### DISCUSSION THREAD

**Discussion by ediskandarov-cur** [sqlalchemy/sqlalchemy#7157] [Usage Questions] [answered]
Hi there,

TLDR:

I was debugging these logs. Asyncio loop works in debug mode and emit logs when the loop was blocked.
```
Oct 7 06:49:28 PM  2021-10-07 13:49:28.513 | DEBUG    | foobar.crud:get_last_known_obj:409 - Statement ready. Sleeping...
Oct 7 06:49:28 PM  Executing <Task pending name='Task-1026' coro=<RequestResponseCycle.run_asgi() running at /app/.venv/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py:373> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f2e897d8580>()] created at /usr/local/lib/python3.8/asyncio/base_events.py:422> cb=[set.discard()] created at /app/.venv/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py:206> took 0.485 seconds
Oct 7 06:49:28 PM  2021-10-07 13:49:28.614 | DEBUG    | foobar.crud:get_last_known_obj:411 - Executing...

Oct 7 06:49:36 PM  2021-10-07 13:49:36.031 | DEBUG    | foobar.crud:get_last_known_obj:413 - Execution finished. Sleeping...
Oct 7 06:49:36 PM  Executing <Task pending name='Task-1026' coro=<RequestResponseCycle.run_asgi() running at /app/.venv/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py:373> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f2e898d1040>()] created at /usr/local/lib/python3.8/asyncio/base_events.py:422> cb=[set.discard()] created at /app/.venv/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py:206> took 0.415 seconds
Oct 7 06:49:36 PM  2021-10-07 13:49:36.133 | DEBUG    | foobar.crud:get_last_known_obj:417 - Getting results
```

The code causing these logs:
```python
async def get_last_known_obj(db: AsyncSession) -> Optional[models.Entity]:
    logger.debug("Building statement for get last known obj")
    stmt = select(models.Entity).order_by(models.Entity.number.desc())
    logger.debug("Statement ready. Sleeping...")
    await asyncio.sleep(0.1)
    logger.debug("Executing...")
    result = await db.execute(stmt)
    logger.debug("Execution finished. Sleeping...")

    await asyncio.sleep(0.1)

    logger.debug("Getting results")

    obj = result.scalars().first()
    logger.debug("Result is ready")

    return obj
```

I'm using

```
python==3.8.12
asyncpg==0.24.0
sqlalchemy==1.4.25
```

create engine a regular way:

```python
engine = create_async_engine(
    db_url,
    echo=settings.sqlalchemy_echo,
    future=True,
)

SessionLocal = sessionmaker(
    engine, expire_on_commit=False, class_=AsyncSession, future=True
)
```

---

What I eventually found is that query generates and executes the following SQL snippet:

```sql
SELECT entity.id, entity.number
FROM entity ORDER BY entity.number DESC
```

That SQL fetches all the records from DB and then inside the application by calling `first`, it picks the first.
But due to a large number of records, the freeze happens.

### Fix

When I explicitly added `.limit(1)` to the statement - the issue has gone.

---

Is it a bug? Shall I explicitly add `.limit(1)` every time I want to use `.first()`?

**Comment by CaselIT** [comment, marked as answer]
Hi,

I think this is working as documented: https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=first#sqlalchemy.engine.Result.first

Note that query in the example above does not generate the sql you listed: it uses desc as order and applies no for update.
But it's similar, since it does not add any limit

**Reply by ediskandarov-cur** [reply]
Thanks for your reply.

> Hi,
> 
> I think this is working as documented: https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=first#sqlalchemy.engine.Result.first

Agree. Then [AsyncResult.first](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncResult.first) needs documentation update too.

> Note that query in the example above does not generate the sql you listed: it uses desc as order and applies no for update. But it's similar, since it does not add any limit

ah, right. Sorry. In my code - there are two similar crud operations. I updated both but took the log from another one.

**Reply by CaselIT** [reply]
> Agree. Then AsyncResult.first needs documentation update too.

indeed. did you find any other place where the documentation is outdated?

**Reply by ediskandarov-cur** [reply]
it seems in https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html

there are many references to [sqlalchemy.engine.Result](https://docs.sqlalchemy.org/en/14/core/connections.html#sqlalchemy.engine.Result)
while it should be pointing to [sqlalchemy.ext.asyncio.AsyncResult](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncResult)

**Reply by zzzeek** [reply]
can you show me one?  the asyncio extension includes methods that return plain Result objects and other methods that return AsyncResult objects.  The Result object is much more prominent as the AsyncResult is not often practical.

**Reply by zzzeek** [reply]
> > Agree. Then AsyncResult.first needs documentation update too.
> 
> indeed. did you find any other place where the documentation is outdated?

im sorry whats the error in the documentation ?

**Reply by CaselIT** [reply]
it's just missing the note as in https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=first#sqlalchemy.engine.Result.first

**Comment by ediskandarov-cur** [comment]
I wonder, what is the benefit of using `.first()` then? It does not add a limit, which makes it impractical.

If I have to explicitly add `limit(1)`, then these work the similar:

```python
stmt = select(models.Entity).limit(1)
result = await db.execute(stmt)

obj = result.scalars().one_or_none()
# vs
obj = result.scalars().first()
```

I'm still puzzled. `stmt.limit(1) and result.one_or_none()` looks much safer and brings less confusion

**Reply by CaselIT** [reply]
I guess `first` is more useful if you replace `execute` with `stream`.

It's there mainly for convenience.

**Reply by ediskandarov-cur** [reply]
🤔 
How does `stream` and `first` work in combination with `with_for_update`?

Does not sound like a good idea

**Reply by CaselIT** [reply]
Sure, I do not advise using with for update with stream.
But note that with for update should not be used without a limit or where, otherwise the all the table gets locked

**Reply by ediskandarov-cur** [reply]
understood. I learned it the hard way :)

I would say it's(first behaviour) counter-intuitive, especially for devs with Django background

Hopefully, this thread reveals some issues with documentation. Shall I file a ticket?

**Reply by CaselIT** [reply]
> I would say it's(first behaviour) counter-intuitive, especially for devs with Django background

I guess a bit it may be. It's important to remember that the execution is all done in `execute`/`stream` 

I think you can add to this one I've opened https://github.com/sqlalchemy/sqlalchemy/issues/7158
